File: python/codegeneration/cmake/codegenerator.py
import sys
from pathlib import Path
sys.path.append(str(Path(__file__).parent.resolve()) + "/../../site-packages/")

from pathlib import Path
import jinja2

# Template files are read globally so it's contents can be shared between calls
with open(Path(__file__).parent.resolve() / 'cmakelists_root.template', 'r') as f:
    CMAKELISTS_ROOT_TEMPLATE = f.read()
with open(Path(__file__).parent.resolve() / 'cmakelists_pkg_group.template', 'r') as f:
    CMAKELISTS_PKG_GROUP_TEMPLATE = f.read()
with open(Path(__file__).parent.resolve() / 'cmakelists_pkg.template', 'r') as f:
    CMAKELISTS_PKG_TEMPLATE = f.read()
with open(Path(__file__).parent.resolve() / 'headerfile.template', 'r') as f:
    HEADER_TEMPLATE = f.read()
with open(Path(__file__).parent.resolve() / 'sourcefile.template', 'r') as f:
    SOURCE_TEMPLATE = f.read()


def _writeTemplateFile(template, build_data, output_file):
    '''
    Helper function for writing contents from a given template file
    '''
    with open(Path(output_file), 'w') as outfile:
        outfile.write(jinja2.Template(template).render(build_data))


def buildPkgGroup(pkg_group, output_dir, user_ctx):
    '''
    Ensures the given package group exists in the filesystem
    '''
    pkg_group_dir = f'{output_dir}/{pkg_group.name()}'
    Path(pkg_group_dir).mkdir(parents=True, exist_ok=True)

    user_ctx['root_entities'].append(pkg_group)

    build_data = {
        'pkgs': pkg_group.children(),
    }

    _writeTemplateFile(CMAKELISTS_PKG_GROUP_TEMPLATE, build_data, f'{pkg_group_dir}/CMakeLists.txt')


def buildPkg(pkg, output_dir, user_ctx):
    '''
    Ensures the given package exists in the filesystem and create its basic structure
    '''
    pkg_group = pkg.parent()
    if pkg_group:
        pkg_dir = f'{output_dir}/{pkg_group.name()}/{pkg.name()}'
    else:
        pkg_dir = f'{output_dir}/{pkg.name()}'
    Path(pkg_dir).mkdir(parents=True, exist_ok=True)

    build_data = {
        'pkg': pkg,
        'deps': [
            dep for dep in pkg.forwardDependencies()
            if not _isNonLakosianDependency(dep)
        ],
    }

    _writeTemplateFile(CMAKELISTS_PKG_TEMPLATE, build_data, f'{pkg_dir}/CMakeLists.txt')
    if pkg_group is None:
        user_ctx['root_entities'].append(pkg)


def _isNonLakosianDependency(entity):
    if entity.type() == pycgn.DiagramType.PackageGroup:
        return entity.name() == 'non-lakosian group'
    if entity.parent():
        return _isNonLakosianDependency(entity.parent())
    return False


def buildComponent(component, output_dir, user_ctx):
    '''
    Build the C++ component in the filesystem with header, source and test files
    '''
    pkg = component.parent()
    assert pkg, f"Package for entity {component.name()} not found."
    pkg_group = pkg.parent()
    if pkg_group:
        pkg_dir = f'{output_dir}/{pkg_group.name()}/{pkg.name()}'
    else:
        pkg_dir = f'{output_dir}/{pkg.name()}'
    component_basename = f'{pkg_dir}/{component.name()}'

    build_data = {
        'component_name': component.name(),
        'package_name': pkg.name(),
        'component_fwd_dependencies': [
            dep.name() for dep in component.forwardDependencies()
            if not _isNonLakosianDependency(dep)
        ],
        'should_generate_namespace': pkg.name().isalnum(),
    }

    _writeTemplateFile(HEADER_TEMPLATE, build_data, component_basename + '.h')
    _writeTemplateFile(SOURCE_TEMPLATE, build_data, component_basename + '.cpp')


def beforeProcessEntities(output_dir, user_ctx):
    user_ctx['root_entities'] = []


def buildPhysicalEntity(pycgn_module, entity, output_dir, user_ctx):
    '''
    API called from DiagramViewer software.
    See software docs for details.
    '''
    global pycgn
    pycgn = pycgn_module

    BUILD_DISPATCH = {
        pycgn.DiagramType.PackageGroup: buildPkgGroup,
        pycgn.DiagramType.Package: buildPkg,
        pycgn.DiagramType.Component: buildComponent
    }
    run = BUILD_DISPATCH.get(entity.type(), lambda _1, _2, _3: None)
    run(entity, output_dir, user_ctx)


def afterProcessEntities(output_dir, user_ctx):
    build_data = {
        'root_entities': user_ctx['root_entities'],
    }

    _writeTemplateFile(CMAKELISTS_ROOT_TEMPLATE, build_data, f'{output_dir}/CMakeLists.txt')

File: python/codegeneration/cmake/codegenerator.js
var FileHandler = new FileIO();

var alphaNumericRegExp = /^[A-Za-z0-9]+$/;
var CMAKELISTS_ROOT_TEMPLATE = ScriptFolder + "/cmakelists_root.template";
var CMAKELISTS_PKG_GROUP_TEMPLATE = ScriptFolder + "/cmakelists_pkg_group.template";
var CMAKELISTS_PKG_TEMPLATE = ScriptFolder + "/cmakelists_pkg.template";
var HEADER_TEMPLATE = ScriptFolder + "/headerfile.template";
var SOURCE_TEMPLATE = ScriptFolder + "/sourcefile.template";
var rootEntities = new Array();

function _writeTemplateFile(template, build_data, output_file) {
    let contents = FileHandler.render(template, build_data);
    if (FileHandler.hasError()) {
        console.log(FileHandler.errorString())
        return;
    }

    FileHandler.saveFile(output_file, contents);
    if (FileHandler.hasError()) {
        console.log(FileHandler.errorString())
        return;
    }
    console.log("Write Template:", output_file);
}

function buildPkgGroup(pkg_group, output_dir, user_ctx) {
    FileHandler.createPath(output_dir + "/" + pkg_group.name)
    rootEntities.push(pkg_group)

    var build_data = {
        'pkgs': pkg_group.children,
    }

    var pkg_group_dir = output_dir + "/" + pkg_group.name;
    var outputFile = pkg_group_dir + "/CMakeLists.txt";
    _writeTemplateFile(CMAKELISTS_PKG_GROUP_TEMPLATE, build_data, outputFile)
}

function buildPkg(pkg, output_dir, user_ctx) {
    var pkg_group = pkg.parent;
    var pkg_dir = "";

    if (pkg_group) {
        pkg_dir += output_dir + "/" + pkg_group.name + "/" + pkg.name;
    } else {
        pkg_dir += output_dir + "/" + pkg.name;
    }
    FileHandler.createPath(pkg_dir)

    let build_data = {
        'pkg': pkg,
        'deps': _getNonLakosianEntities(pkg.fwdDependencies),
    }

    let outputFile = pkg_dir + "/CMakeLists.txt";
    _writeTemplateFile(CMAKELISTS_PKG_TEMPLATE, build_data, outputFile)
    if (pkg_group == null) {
        rootEntities.push(pkg);
    }
}

function _getNonLakosianEntities(entityArray) {
    let ret = new Array();
    for (const dep of entityArray) {
        if (_isNonLakosianDependency(dep)) {
            ret.push(dep);
        }
    }
    return ret;
}

function _isNonLakosianDependency(entity) {
    if (entity.type == "PackageGroup") {
        return entity.name == 'non-lakosian group';
    }
    if (entity.parent) {
        return _isNonLakosianDependency(entity.parent);
    }
    return false;
}

/* JS Lacks isAlphaNumeric, so let's add it. */
function isAlphaNumeric(Value) {
    return (Value.match(alphaNumericRegExp));
}

function buildComponent(component, output_dir, user_ctx) {
    var pkg = component.parent
    var pkg_group = pkg.parent
    var pkg_dir = output_dir;
    if (pkg_group) {
        pkg_dir += "/" + pkg_group.name + "/" + pkg.name;
    } else {
        pkg_dir += "/" + pkg.name;
    }

    var component_basename = pkg_dir + "/" + component.name;

    var build_data = {
        'component_name': component.name,
        'package_name': pkg.name,
        'component_fwd_dependencies': _getNonLakosianEntities(component.fwdDependencies),
        'should_generate_namespace': isAlphaNumeric(pkg.name),
    };

    _writeTemplateFile(HEADER_TEMPLATE, build_data, component_basename + '.h');
    _writeTemplateFile(SOURCE_TEMPLATE, build_data, component_basename + '.cpp');
}

export function beforeProcessEntities(output_dir, user_ctx) {
    rootEntities = new Array();
}

export function buildPhysicalEntity(entity, output_dir, user_ctx) {
    switch(entity.type){
        case "PackageGroup":
            buildPkgGroup(entity, output_dir);
        break;
        case "Package":
            buildPkg(entity, output_dir);
        break;
        case "Component":
            buildComponent(entity, output_dir);
        break;
    }
}

export function afterProcessEntities(output_dir, user_ctx) {
    var build_data = {
        'root_entities': rootEntities
    };

    var outputFile = output_dir + "/CMakeLists.txt";
    _writeTemplateFile(CMAKELISTS_ROOT_TEMPLATE, build_data, outputFile)
}

